Fix ccs permission for search with a scroll id

ywangd opened the pull request:
CCS with remote indices only does not require any privileges on the local cluster.
This PR ensures that search with scroll follow the permission model.

Resolves: #61828

ywangd committed:
Fix ccs permission for scroll search

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authorization)

ywangd committed:
Checkstyle

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-61828-ccs-scroll-privileges

ywangd committed:
Fix test failure

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchHelper.java:
```diff
@@ -108,6 +109,10 @@ static ParsedScrollId parseScrollId(String scrollId) {
     }
 
+    static boolean hasLocalIndices(String scrollId) {
+        return Arrays.stream(parseScrollId(scrollId).getContext()).anyMatch(c -> c.getClusterAlias() == null);
```
I would suggest that `hasLocalIndices` be a method of `ParsedScrollId` , it's more OO this way.

ywangd replied:
You are absolutely right. Now I wonder why I didn't do it initially ...

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -261,7 +262,7 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
                 // index and if they cannot, we can fail the request early before we allow the execution of the action and in
                 // turn the shard actions
-                if (SearchScrollAction.NAME.equals(action)) {
+                if (SearchScrollAction.NAME.equals(action) && ((SearchScrollRequest) request).hasLocalIndices()) {
```
I would avoid throwing exceptions in this part of the code, it's very brittle, better call the `listener.onFailure` .

ywangd replied:
This is a good catch. Thanks. I have added the exception handling and listener call. It is a pretty basic `try..catch` block. Please let me know if there are better way or convention to do it differently. Thanks.

albertzaharovits reviewed (changes requested):
Besides the two minor points, can you also please add a test similar to `AuthorizationServiceTests#testUserWithNoRolesCanPerformRemoteSearch` ?

ywangd committed:
Address feedback

ywangd commented:
> Besides the two minor points, can you also please add a test similar to `AuthorizationServiceTests#testUserWithNoRolesCanPerformRemoteSearch` ?
>
I added a new [test](https://github.com/elastic/elasticsearch/pull/62053/files#diff-3dec0fec51a395d9a9247e011cc2808dR415). Thanks.

ywangd committed:
Add randomness to test

ywangd commented:
> I added a new [test](https://github.com/elastic/elasticsearch/pull/62053/files#diff-3dec0fec51a395d9a9247e011cc2808dR415).
>
Updated the test to add randomness.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-61828-ccs-scroll-privileges

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-61828-ccs-scroll-privileges

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java:
```diff
@@ -411,6 +412,31 @@ public void testUserWithNoRolesCanPerformRemoteSearch() throws IOException {
 
+    public void testUserWithNoRolesPerformsRemoteSearchWithScroll() throws IOException {
+        final ParsedScrollId parsedScrollId = mock(ParsedScrollId.class);
+        final boolean hasLocalIndices = randomBoolean();
```
I would've made this variable take both values in this test, i.e. don't rely on randomisation.

In principle, it's better to avoid randomisation for test coverage, and use it for variables that don't have any bearing on the output of the test. That's in principle, but sometimes the input value space is large and randomisation might be used on the independent variables as well (more often inside, but sometimes across, value spaces that change the outcome of the test).

ywangd replied:
Thanks that's a good point. I changed it to always test both situations.

albertzaharovits commented on server/src/test/java/org/elasticsearch/action/search/ParsedScrollIdTests.java:
```diff
@@ -0,0 +1,42 @@
+        boolean hasLocal = false;
+        for (int i = 0; i < nResults; i++) {
+
+            String clusterAlias = randomBoolean() ? randomAlphaOfLength(8) : null;
```
ditto https://github.com/elastic/elasticsearch/pull/62053/files#r487806064

ywangd replied:
I tried to remove the randomness, but then decided to keep it as is. Here is why: The intention is to test that "any local indices will be reported as hasLocalIndices". The "any" bit makes it more suitable for randomness, because it can be:
* A single search context
* Multiple search contexts where 
  * the local one can be in the first position, second position, last position etc. 
  * Or there might be more than one local contexts, e.g. all of them are local contexts or only one of them is not local context.

The randomness in the test is intended to cover all above situations. If removed, the coverage will be reduced (or we can replace it with a very bloated test method, but I think that is less ideal).

albertzaharovits commented on server/src/test/java/org/elasticsearch/action/search/ParsedScrollIdTests.java:
```diff
@@ -0,0 +1,42 @@
+import org.elasticsearch.test.ESTestCase;
+
+public class ParsedScrollIdTests extends ESTestCase {
+    public void testHasLocalIndices() {
```
This test relies on the `TransportSearchHelperTests#testParseScrollId` , where, given `SearchShardTargets`, the scroll id is encoded and then decoded, and the test asserts that the "cluster alias" shows up after decoding.

IMO, this test adds little value on top of that, and it's coded away from the above mentioned piece of code.
My personal approach would've been to parametrise the `TransportSearchHelperTests#generateQueryResults` method to allow it to generate scroll ids with/without cluster aliases, and assert the `ParsedScrollId#hasLocalIndices` method there as well (probably in another test but not necessarily so).

ywangd replied:
I am not sure if I follow your comment. This test is an unit test for `ParsedScrollId` and has nothing to do with `TransportSearchHelperTests#testParseScrollId `. Maybe you were commenting on a previous version of this test, i.e. the one before I moved this method as part of `ParsedScrollId` following your comment of [OO design](https://github.com/elastic/elasticsearch/pull/62053#discussion_r484745362)?

albertzaharovits commented on server/src/main/java/org/elasticsearch/action/search/SearchScrollRequest.java:
```diff
@@ -83,6 +83,10 @@ public SearchScrollRequest scrollId(String scrollId) {
     }
 
+    public ParsedScrollId parsedScrollId() {
+        return TransportSearchHelper.parseScrollId(scrollId);
```
In principle, I would shy away from getters that do any processing, unless the method has a verb in the name.

ywangd replied:
Given `TransportSearchHelper.parseScrollId(scrollId)` can throw exception, I think it is fair to name it with a verb. I updated to be `parseScrollId` (same name as `TransportSearchHelper#parseScrollId`). The extra layer of proxy (from ParsedScrollId to TransportSearchHelper) is because `TransportSearchHelper` is not public and it feels better to keep it that way.

albertzaharovits reviewed (approved):
LGTM
I've made some "In principle" comments in order to expose my personal thinking, but I'm happy to merge this as is.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -262,7 +264,16 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
                 // turn the shard actions
                 if (SearchScrollAction.NAME.equals(action)) {
-                    authorizeIndexActionName(action, authorizationInfo, null, listener);
+                    try {
```
As you've mentioned the try-catch structure unfortunately doesn't fit too nicely with the async listener idiom .
I personally would've maintained a check like `request instanceof SearchScrollRequest && ((SearchScrollRequest)request).scrollId() != null` as I'm not a fan of catching runtime exceptions . 

Again, this is a feedback with my perspective, the codebase is cot consistent in this regard; it's your choice.

ywangd replied:
I changed it to use `ActionRunnable.supply(...).run()`, which takes a `CheckedSupplier` to handle the exception. Honestly, the code is less readable with this change. But it is more in line with the overwhelming callback style.

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-61828-ccs-scroll-privileges

ywangd committed:
Address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into es-61828-ccs-scroll-privileges
